File: java/repository/src/main/java/database/RallyDatabaseRepository.java
package database;

import exceptions.DatabaseException;
import exceptions.NotImplementedMethodException;
import interfaces.RallyRepository;
import models.Rally;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import utils.JdbcUtils;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Properties;
import java.util.function.Function;
import java.util.function.Supplier;

public class RallyDatabaseRepository implements RallyRepository {

    private JdbcUtils jdbcUtils;

    private static final String INSERT_RALLY_SQL_STRING = "insert into rallies(engine_capacity) values (?)";
    private static final String FIND_ALL_RALLIES_SQL_STRING = "select rally_id, engine_capacity, number_of_participants from rallies";
    private static final String FIND_BY_ENGINE_CAPACITY_SQL_STRING = "select rally_id, engine_capacity, number_of_participants from rallies where engine_capacity=?";

    private Logger logger = LogManager.getLogger();

    public RallyDatabaseRepository(Properties databaseConnectionProperties) {
        logger.info("Creating rally database repository...");
        this.jdbcUtils = new JdbcUtils(databaseConnectionProperties);
        logger.info("Rally database repository created");
    }

    @Override
    public Rally save(Rally model) {
        Function<Rally, Rally> saveRallyFunction = rally -> {

            logger.traceEntry("Saving rally {}", model);

            var existingRallyOptional = findRallyByEngineCapacity(model.getEngineCapacity());
            if (existingRallyOptional.isPresent()) {
                logger.info("Rally {} already exists", existingRallyOptional.get());
                logger.traceExit();
                return existingRallyOptional.get();
            }

            try (Connection connection = jdbcUtils.getConnection();
                 PreparedStatement insertRallyPreparedStatement =
                         connection.prepareStatement(INSERT_RALLY_SQL_STRING, Statement.RETURN_GENERATED_KEYS)) {

                insertRallyPreparedStatement.setInt(1, model.getEngineCapacity());
                insertRallyPreparedStatement.executeUpdate();
                try (ResultSet generatedKeys = insertRallyPreparedStatement.getGeneratedKeys()) {

                    if (generatedKeys.next()) {
                        Long generatedId = generatedKeys.getLong(1);
                        Rally rallyWithId = new Rally(model);
                        rallyWithId.setId(generatedId);
                        logger.info("New rally created {}", rallyWithId);
                        logger.traceExit();
                        return rallyWithId;
                    }
                }
            } catch (SQLException e) {
                logger.error(e);
            }
            logger.traceExit();
            return null;
        };

        var savedRally = saveRallyFunction.apply(model);

        if (savedRally == null)
            throw new DatabaseException("DatabaseException occurred while saving rally " + model);

        return savedRally;
    }


    @Override
    public Iterable<Rally> findAll() {
        Supplier<Iterable<Rally>> findAllSupplier = () -> {
                logger.traceEntry("Returning all rallies...");

                try (Connection connection = jdbcUtils.getConnection();
                     PreparedStatement findAllRalliesPreparedStatement = connection.prepareStatement(FIND_ALL_RALLIES_SQL_STRING)) {
                    try (ResultSet foundRallies = findAllRalliesPreparedStatement.executeQuery()) {
                        List<Rally> allRallies = new ArrayList<>();
                        while (foundRallies.next()) {
                            Long rallyId = foundRallies.getLong("rally_id");
                            int engineCapacity = foundRallies.getInt("engine_capacity");
                            int numberOfParticipants = foundRallies.getInt("number_of_participants");
                            Rally currentRally = new Rally(engineCapacity, numberOfParticipants);
                            currentRally.setId(rallyId);
                            allRallies.add(currentRally);
                        }
                        logger.traceExit();
                        return allRallies;
                    }
                } catch (SQLException exception) {
                    logger.error(exception);
                }
                logger.traceExit();
                return null;
        };

        var rallies = findAllSupplier.get();

        if (rallies == null)
            throw new DatabaseException("DatabaseException occurred while returning all rallies");

        return rallies;
    }



    @Override
    public Optional<Rally> findRallyByEngineCapacity(int engineCapacity) {
        Function<Integer, Optional<Rally>> findRallyByEngineCapacityFunction = capacity -> {
            logger.traceEntry("Finding one rally by engine capacity...");

            try (Connection connection = jdbcUtils.getConnection();
                 PreparedStatement findByEngineCapacityPreparedStatement = connection.prepareStatement(FIND_BY_ENGINE_CAPACITY_SQL_STRING)) {
                findByEngineCapacityPreparedStatement.setInt(1, engineCapacity);
                try (ResultSet resultSet = findByEngineCapacityPreparedStatement.executeQuery()) {
                    if (resultSet.next()) {
                        Long rallyId = resultSet.getLong("rally_id");
                        int existingEngineCapacity = resultSet.getInt("engine_capacity");
                        int numberOfParticipants = resultSet.getInt("number_of_participants");
                        var existingRally = new Rally(existingEngineCapacity, numberOfParticipants);
                        existingRally.setId(rallyId);
                        logger.info("Found rally {}", existingRally);
                        logger.traceExit();
                        return Optional.of(existingRally);
                    } else
                        return Optional.empty();
                }
            } catch (SQLException exception) {
                logger.error(exception);
            }
            logger.traceExit();
            return null;
        };

        var foundRally = findRallyByEngineCapacityFunction.apply(engineCapacity);

        if (foundRally == null)
            throw new DatabaseException("DatabaseException occurred while searching for rally with engineCapacity " + engineCapacity);

        return foundRally;
    }

    @Override
    public Rally findOne(Long modelID) {
        throw new NotImplementedMethodException();
    }
}

File: csharp/Repository/Database/RallyDatabaseRepository.cs
using DatabaseConnectionFactories;
using Domain.Models;
using log4net;
using Repository.Interfaces;
using Repository.Utils;
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Common;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Repository.Database
{
    public class RallyDatabaseRepository : IRallyRepository
    {
        private DatabaseUtils utils;
        private static readonly ILog log = LogManager.GetLogger("DallyDatabaseRepository");

        private static readonly String INSERT_RALLY_SQL_STRING = "insert into rallies(engine_capacity) values (@engine_capacity_param)";
        private static readonly String FIND_ALL_RALLIES_SQL_STRING = "select rally_id, engine_capacity, number_of_participants from rallies";
        private static readonly String FIND_BY_ENGINE_CAPACITY_SQL_STRING = "select rally_id, engine_capacity, number_of_participants from rallies where engine_capacity=@engine_capacity_param;";

        public RallyDatabaseRepository(IDictionary<String, String> databaseConnectionProperties)
        {
            log.Info("Creating rally database repository...");
            this.utils = new DatabaseUtils(databaseConnectionProperties);
            log.Info("Rally database repository created!");
        }
        public Rally? GetByEngineCapacity(int engineCapacity)
        {
            log.Info("Searching for rally by engine capacity " + engineCapacity);

            try
            {
                using (var connection = utils.GetConnection())
                {
                    using (var comamnd = connection.CreateCommand())
                    {
                        comamnd.CommandText = FIND_BY_ENGINE_CAPACITY_SQL_STRING;

                        var engineCapacityParam = comamnd.CreateParameter();
                        engineCapacityParam.ParameterName = "@engine_capacity_param";
                        engineCapacityParam.Value = engineCapacity;

                        comamnd.Parameters.Add(engineCapacityParam);
                        using (var reader = comamnd.ExecuteReader())
                        {
                            if (reader.Read())
                            {
                                return GetCurrentRallyFromReader(reader);
                            }
                            return null;
                        }
                    }
                }
            }
            catch(DbException ex)
            {
                log.Error(ex.Message, ex);
                throw new DatabaseException("Error occured while searching for rally with engine capacity " + engineCapacity);
            }
        }
        public Rally Add(Rally model)
        {
            log.Info("Adding new rally " + model.ToString());
            var existingRally = this.GetByEngineCapacity(model.EngineCapacity);
            if (existingRally != null)
                return existingRally;

            try
            {
                using (var connection = utils.GetConnection())
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = INSERT_RALLY_SQL_STRING;

                    var engineCapacityParameter = command.CreateParameter();
                    engineCapacityParameter.ParameterName = "@engine_capacity_param";
                    engineCapacityParameter.Value = model.EngineCapacity;
                    command.Parameters.Add(engineCapacityParameter);
                    command.ExecuteNonQuery();
                    Rally? savedRally = GetByEngineCapacity(model.EngineCapacity);

                    if (savedRally == null)
                        throw new DatabaseException("Error occured while saving rally " + model);
                    return savedRally;
                }
            }
            catch(DbException ex)
            {
                log.Error(ex.Message, ex);
                throw new DatabaseException("Error occured while saving rally " + model);
            }
        }

        public ICollection<Rally> GetAll()
        {
            log.Info("Returning all rallies...");

            try
            {
                using (var connection = utils.GetConnection())
                {
                    using (var command = connection.CreateCommand())
                    {
                        command.CommandText = FIND_ALL_RALLIES_SQL_STRING;
                        using (var reader = command.ExecuteReader())
                        {
                            ICollection<Rally> rallies = new List<Rally>();
                            while (reader.Read())
                            {
                                rallies.Add(GetCurrentRallyFromReader(reader));
                            }
                            return rallies; 
                        }
                    }
                }
            }
            catch(DbException exception)
            {
                log.Error(exception.Message, exception);
                throw new DatabaseException("Error occured while searching for all rallies");
            }

        }

        private Rally GetCurrentRallyFromReader(IDataReader reader)
        {
            long rallyId = reader.GetInt64(0);
            int exitingEngineCapacity = reader.GetInt32(1);
            int noOfParticipants = reader.GetInt32(2);
            Rally rally = new Rally(exitingEngineCapacity, noOfParticipants);
            rally.Id = rallyId;
            return rally;
        }



        public Rally GetById(long rallyId)
        {
            throw new NotImplementedException();
        }

    }
}
